src/community/greedy_modularity.jl

```diff
@@ -0,0 +1,101 @@
+function community_detection_greedy_modularity(g::AbstractGraph; weights::AbstractMatrix=weights(g))
+    if is_directed(g)
+        throw(ArgumentError("The graph must not be directed"))
+    end
+    n = nv(g)
+    c = Vector{Int}(1:n)
+    cs = Vector{Vector{Int}}(undef, n)
+    T = float(eltype(weights))
+    qs = Vector{T}(undef, n)
+    Q, e, a = compute_modularity(g, c, weights)
+    m = sum(a)
+    cs[1] = copy(c)
+    qs[1] = Q
+    for i in 2:n
+        Q = modularity_greedy_step!(g, Q, e, a, c, m)
+        cs[i] = copy(c)
+        qs[i] = Q
+    end
+    imax = argmax(qs)
+    return rewrite_class_ids(cs[imax])
+end
+
+function modularity_greedy_step!(
+    g::AbstractGraph,
+    Q::T,
+    e::AbstractMatrix{T},
+    a::AbstractVector{T},
+    c::AbstractVector{<:Integer},
+    m::T
+) where {T}
+    n = nv(g)
+    dq_max::typeof(Q) = typemin(Q)
+    to_merge::Tuple{Int,Int} = (0, 0)
+    for edge in edges(g)
```

Comment:
investigate the case of self-loops

Reply:
Self-loops indeed appear in modularity computation (in a correct way)
They do participate in modularity optimization step and they can also impact which merge is optimal. Yet we never merge a cluster with itself as we check that ends of an edge belong to different clusters at traversal stage.